new postgresql_query_module

archf opened the pull request:
##### SUMMARY
New `postgresql_query` module as a companion to other `postgresql_*` modules. Run queries passed as arg or as sql scripts. See documentation in module.

I know of [PR #20793](https://github.com/ansible/ansible/pull/20793) but it seems stuck. I made this one in a attempt to solve the issues and add the capabilities to execute a .sql script file. If someone can review i will resolve any upcoming issues.

I adapted from the one embedded i use in [this role i made](https://github.com/archf/ansible-sql-runner). If anyone shows interests i can also make a PR for the impala_query module.

##### ISSUE TYPE

 - New Module Pull Request

archf committed:
new postgresql_query_module

Signed-off-by: Félix Archambault <fel.archambault@gmail.com>

Akasurde commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,341 @@
+     Number of rows affected are returned as "row_count".
+   - Can read queries from a .sql script files
+   - SQL scripts may be templated with ansible variables at execution time
+version_added: "2.3"
```
s/2.3/2.4/

archf replied:
fixed

Akasurde commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,341 @@
+DOCUMENTATION = '''
+---
+module: postgresql_query
+short_description: Run arbitrary queries on a postgresql database.
```
s/postgresql/PostgreSQL/g

Akasurde commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,341 @@
+    changed = False
+
+    # set changed flag only on non read-only command
+    if "UPDATDE" in statusmessage or "INSERT" in statusmessage or \
```
IMHO this should be `UPDATE`

archf replied:
fixed, good catch, tx

gundalow commented:
```
lib/ansible/modules/database/postgresql/postgresql_query.py:0:0: E305 DOCUMENTATION.options.fact: expected a dictionary for dictionary value @ data['options']['fact']. Got 'Append the query_results to this key value, making this new variable\navailable to subsequent plays during an ansible-playbook run.\n'
lib/ansible/modules/database/postgresql/postgresql_query.py:0:0: E307 version_added should be 2.4. Currently 2.3
lib/ansible/modules/database/postgresql/postgresql_query.py:0:0: E316 ANSIBLE_METADATA.metadata_version: required key not provided @ data['metadata_version']. Got None
lib/ansible/modules/database/postgresql/postgresql_query.py:0:0: E316 ANSIBLE_METADATA.version: extra keys not allowed @ data['version']. Got '1.0'
lib/ansible/modules/database/postgresql/postgresql_query.py:0:0: E319 RETURN.ansible_facts.description: required key not provided @ data['description']. Got None
lib/ansible/modules/database/postgresql/postgresql_query.py:0:0: E319 RETURN.ansible_facts.my_key: extra keys not allowed @ data['my_key']. Got {'column2': 'value', 'column1': 'value'}
lib/ansible/modules/database/postgresql/postgresql_query.py:0:0: E319 RETURN.ansible_facts.returned: required key not provided @ data['returned']. Got None
lib/ansible/modules/database/postgresql/postgresql_query.py:0:0: E319 RETURN.ansible_facts.type: required key not provided @ data['type']. Got None
lib/ansible/modules/database/postgresql/postgresql_query.py:338:0: E107 Imports should be directly below DOCUMENTATION/EXAMPLES/RETURN/ANSIBLE_METADATA.
 ```
http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html should assist you in solving the above

archf committed:
postgresql_query_module: bump version_added and fix typo

archf committed:
postgresql_query_module: group imports below documentation

an attempt to fix E107

archf commented:
@gundalow 

does this work for you from my branch?

```
archf@laptop ~/r/ansible postgresql_query_module!> ./test/sanity/validate-modules/validate-modules --arg-spec --warnings  lib/ansible/modules/database/postgresql/postgresql_query.py
Traceback (most recent call last):
  File "./test/sanity/validate-modules/validate-modules", line 40, in <module>
    from ansible.utils.plugin_docs import BLACKLIST, get_docstring
ImportError: No module named plugin_docs
archf@laptop ~/r/ansible postgresql_query_module!> find lib/ansible/utils -name \*plugin_docs\*
lib/ansible/utils/plugin_docs.py
```

I did `source hacking/env-setup` just before...

archf committed:
postgresql_query_module: fix: metadata and statusmessage checks

archf committed:
fix documentation issues

Signed-off-by: Félix Archambault <fel.archambault@gmail.com>

archf commented:
I think i fixed all issues, will see...

archf commented:
Ammended, i shouldn't work so late :)

sivel commented:
Closing and reopening to retrigger CI

archf committed:
fix: exception ordering

put sub class exception clause before its ancestor’s class

Signed-off-by: Félix Archambault <fel.archambault@gmail.com>

gundalow commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,344 @@
+        - If enabled, This WILL run the query in check_mode.
+      required: false
+      default: false
+      choices: [true, false]
```
Please replace with
`type: bool`

archf replied:
fixed

gundalow commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,344 @@
+  named_args:
+    description:
+      - A dictionary of key-value arguments to pass to the query.
+      - Cannot be used with positional_args.
```
`I(positional_args)`

archf replied:
not sure what this means...

archf committed:
docs: improve doc

fix autocommit input arg description
comma at the end of sentences
added Will authorship

Signed-off-by: Félix Archambault <fel.archambault@gmail.com>

wrouesnel commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,344 @@
+  autocommit:
+      description:
+        - Enable transaction autocommit.
+        - If enabled, This WILL run the query in check_mode.
```
I don't love this as behavior you can just "turn on" since it's potentially very surprising. If this is enabled I think the module should stop supporting check_mode (or just lie and always report a change). At the very least there should be a big warning printed about what's going to happen, since presumably users test their playbooks on something non-vital.

wrouesnel commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,344 @@
+'''
+
+EXAMPLES = '''
+# Insert or update a record in a table with positional arguments
```
Comment does not match example.

archf replied:
fixed

wrouesnel commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,344 @@
+    - "positional string value 1"
+    - "positional string value 2"
+
+# Insert or update a record in a table with named arguments
```
Comment does not match example.

archf replied:
fixed

wrouesnel commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,344 @@
+        db_connection = psycopg2.connect(**kw)
+
+        # Enable autocommit on demand only so we can create databases
+        if module.params['autocommit']:
```
This is very non-deterministic if it's going to be supported since psycopg2 will use the default of the server it connects to, which could be anything. If not set, this should be forced to `psycopg2.extensions.ISOLATION_LEVEL_READ_COMMITTED` IMO.

archf replied:
will give a look.

archf replied:
I agree, but i can't find where this would be configured on the server side.
Maybe [here](https://www.postgresql.org/docs/9.5/static/runtime-config-connection.html#RUNTIME-CONFIG-CONNECTION-SETTINGS) or [here](https://www.postgresql.org/docs/9.5/static/runtime-config-client.html)

We could always do something like:

``` python      
        if module.params['autocommit']:
            if psycopg2.__version__ >= '2.4.2':
                db_connection.autocommit = True
            else:
                db_connection.set_isolation_level(psycopg2.extensions.ISOLATION_LEVEL_AUTOCOMMIT)
        else:
            if psycopg2.__version__ >= '2.4.2':
                db_connection.autocommit = False
```

drrtuy replied:
For clarity postgres doesn't support READ UNCOMMITED isolation level so it is not clear what wrousnel meant in his note.

archf replied:
Thanks for the info!

wrouesnel commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,344 @@
+   - Query results are returned as "query_results" as a list of dictionaries.
+     Number of rows affected are returned as "row_count".
+   - Can read queries from a .sql script files.
+   - SQL scripts may be templated with ansible variables at execution time.
```
Should have a note here that check mode will run the query and then roll it back. Something like:
```
- Ansible check mode will run the query and then roll it back. 
  Be careful using long running queries if you're going to use check mode.
```

archf replied:
Will add a note as suggested. What's happening with long queries in check mode?

wrouesnel replied:
If you work on a large table but plan to rollback, on a busy-ish DB you're potentially building up a lot of data in the journal. Which might be fine to do once, but if you're thinking check mode == no effect, could be a surprise.

archf replied:
Ok, will add a notice.

archf replied:
notice added.

wrouesnel reviewed (changes requested):
Found a couple of issues. I like the check mode support, but it also gives me the creeps because of the potential damage to an unwary user.

One final note I have is whether maybe check_mode should have an alternate mode of operation - i.e. a `check_query` option which runs a totally separate query that is expected to report whether something would be done. This could be made required if using the `autocommit` option.

I really dislike the `autocommit` option - I feel like creating databases is something that should be out of scope for this module because there's already the postgresql_db module - modifying an `*.sql` file doesn't seem like too big an ask when automating.

wrouesnel commented:
Another issue, this stanza from over in my original PR #20793 is more or less necessary to handle any schema with timestamps in it:
```python
    # Setup a datetime to string case
    def cast_timestamp(value, cur):
        if value is None:
            return None
        return str(value)

    cursor.execute("SELECT NULL::timestamp")
    oid = cursor.description[0][1]
    sqltype = psycopg2.extensions.new_type((oid,), "timestamp", cast_timestamp)
    psycopg2.extensions.register_type(sqltype)
```
There are probably some more that should be handled, but this one handles an obvious case and presumably we just extend it if we find more canonical uses cases.

archf committed:
postgresql_query_module: fix examples in doc

description not matching args
wrong args

archf committed:
update module description

Signed-off-by: Félix Archambault <fel.archambault@gmail.com>

archf commented:
Thanks for your review. Regarding the autocommit I think however having versionned and self-contained `.sql` scripts is valuable. A developer might develop scripts outside of `ansible` before passing them to ops or devops guys for deployment with ansible. It also allows to deploy entire apps from simple scripts sorted in alphabetical order on filesystem.

 For sure using `postgresql_db` is the ansible way to do it. `postgresql_db` enables autocommit. I don't remember how they handles check_mode. They may have the same shortcomings.

Autocommit is not the default and for sure subsequent tasks in `check_mode` would fail if skipping those. It is an advanced feature for those knowing what they are doing. Tasks using autocommit should be explicitly isolated if they need to be skipped to avoid reporting "changed task"  and that could be achieve with tags or in different play or whatever the mean.

For the timestamps, what's the default behavior? Value is interpreted as an integer instead of string?

wrouesnel commented:
Tripped this bug while testing:
```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: TypeError: __init__() takes exactly 2 arguments (1 given)
fatal: [xen]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_I9p3o6/ansible_module_postgresql_query.py\", line 347, in <module>\n    main()\n  File \"/tmp/ansible_I9p3o6/ansible_module_postgresql_query.py\", line 344, in main\n    rowcount=rowcount)\n  File \"/tmp/ansible_I9p3o6/ansible_modlib.zip/ansible/module_utils/basic.py\", line 2122, in exit_json\n  File \"/tmp/ansible_I9p3o6/ansible_modlib.zip/ansible/module_utils/basic.py\", line 2115, in _return_formatted\n  File \"/tmp/ansible_I9p3o6/ansible_modlib.zip/ansible/module_utils/basic.py\", line 510, in remove_values\n  File \"/tmp/ansible_I9p3o6/ansible_modlib.zip/ansible/module_utils/basic.py\", line 472, in _remove_values_conditions\nTypeError: __init__() takes exactly 2 arguments (1 given)\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 1}
```
This happens when you pass a non-standard structure back to the ansible module - the issue is the Psycopg2 dictionary iterator. You need to iterate it and yield the actual dictionaries.

wrouesnel commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,347 @@
+        # the SQL, so we act consistent and return the empty set when there's
+        # nothing to return.
+        try:
+            query_results = cursor.fetchall()
```
This line needs to be patched to this:
```
query_results = [ dict(row) for row in cursor.fetchall() ]
```
to not brake `module_utils/basic.py` - since it can't cope with complicated stucts (it tries to initialize them when copying in the latest devel head).

archf replied:
Do you have a sample of what your are trying to return? You are fetching a json blob?

wrouesnel replied:
Nothing surprising is being returned - it's just cursor.fetchall() returns an array of RealDictCursor objects, not python dict's which the Ansible module class can't handle - they need to be reified into actual python dictionarys (which the comprehension does).

archf replied:
ok, i fixed it.

archf committed:
postgresql_query_module: update doc: warned about check_mode risks

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,347 @@
+author:
+    - "Felix Archambault (@archf)"
+    - "Will Rouesnel (@wrouesnel)"
+'''
```
Existing documentation should be reused not duplicated. Reuse existing [documentation fragment](http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#documentation-fragments):
```
extends_documentation_fragment:
- postgres
```

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,347 @@
+    description:
+      - name of the database to run queries against.
+    required: false
+    default: postgres
```
Default value for `db` parameter isn't enforced in `argument_spec`: I would rather remove this line or add a default value for `db` in `argument_spec`.

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,347 @@
+  positional_args:
+    description:
+      - A list of values to be passed as positional arguments to the query.
+      - Cannot be used with named_args.
```
Use [formatting function](http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#formatting-options) when referring to other options.

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,347 @@
+  query:
+    description:
+      - SQL query to run. Variables can be escaped with psycopg2 syntax.
+      - Can be a sql scripts file.
```
Another parameter dedicated to SQL script could be used (for example `script`). Then `query` and `script` would be mutually exclusive. The type of `script` would be `path`, the main benefit is that the value would be [expanded](https://github.com/archf/ansible/blob/e7958745d7c9812d0c1cd64de09be92e1f302e09/lib/ansible/module_utils/basic.py#L1801) using `os.path.expanduser` and `os.path.expandvars`.

s/sql/SQL/

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,347 @@
+    default: null
+  query:
+    description:
+      - SQL query to run. Variables can be escaped with psycopg2 syntax.
```
Add a link to Psycopg documentation using `U` formatting function.

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,347 @@
+    db: acme
+    login_user: django
+    login_password: ceec4eif7ya
+    query: SELECT * FROM some_table WHERE a_column=%(a_value)s AND b_column=%(b_value)s
```
Remove the first `query` occurrence.

pilou- replied:
Need to be updated

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,347 @@
+    sample: 5
+'''
+
+HAS_PSYCOPG2 = False
```
`HAS_PSYCOPG2` is already defined in `ansible.module_utils.postgres` which is imported below. Remove this block.

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,347 @@
+def main():
+    argument_spec = pgutils.postgres_common_argument_spec()
+
+    argument_spec.update(dict(
```
default value of `default` is `None`, there is no need to specify `default=None`. default value of `type` is `str`, there is no need to use `type="str"`.

pilou- replied:
Done

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,347 @@
+        supports_check_mode=True,
+    )
+
+    if not HAS_PSYCOPG2:
```
`pgutils.ensure_libs()` is used below, there is no need to check `HAS_PSYCOPG2` here.

archf replied:
fixed

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,347 @@
+        module.fail_json(msg="unable to connect to database: %s" % e, exception=traceback.format_exc())
+
+    except Exception:
+        e = get_exception()
```
`get_exception` was used for Python 2.4 compatibility, support for 2.4 on the managed system's side was dropped: use native Python exception handling.

archf replied:
fixed

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,347 @@
+        else:
+            db_connection.commit()
+
+    db_connection.close()
```
`db_connection.close()` is not called when `cursor.execute` raises an exception.

archf replied:
fixed

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,347 @@
+                         query_arguments=arguments)
+
+    ansible_facts = {}
+    query_results = []
```
`rowcount` could be set to `0` then the `else` block of the conditional statement could be removed.

archf replied:
fixed, this is a good idea

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,347 @@
+    # set changed flag only on non read-only command
+    if ("UPDATE" in statusmessage or "INSERT" in statusmessage or
+            "UPSERT" in statusmessage or "DELETE" in statusmessage):
+        changed = True
```
What about `MOVE`, `COPY`, `CREATE` ? It doesn't seem reliable, `changed` should be `True` if the query has been executed. By the way I am not sure if supporting `check_mode` is a good idea. `command` and `shell` modules don't support `check_mode`, it would be fine it this module didn't support `check_mode`.

archf replied:
You are right, i had some thoughts about this. I added more (see new commit). I dropped check_mode support also as suggested for now.

pilou- reviewed (commented):
I am not sure if this module is useful. What are the main advantages over user 'psql' command ?

Anyway, this module should not support check_mode. Having no test at all is not an option.

wrouesnel commented:
@pilou- this module lets you send parameterized queries to Postgres, and retrieve the data as JSON dictionaries, which makes for easy and consistent access as registered variables or facts within an ansible playbook.

It also fills an obvious gap in the module space - utilities to create databases and users are somewhat pointless without the ability to initialize them, and for applications which use Postgres as a backend you generally want idempotent setup to happen - e.g. the original version of this I wrote to handle configuring auth providers in Gogs while provisioning Postgres databases and users for Gogs.

archf commented:
It's about tighter integration with ansible. Passing values, returning
values...

Else in the spirit of what you said there would be only 3 modules in
ansible: raw, shell and command.

On Jul 19, 2017 9:11 PM, "Pilou" <notifications@github.com> wrote:

> *@pilou-* commented on this pull request.
>
> I am not sure if this module is useful. What are the main advantages over
> user 'psql' command ?
>
> Anyway, this module should not support check_mode. Having no test at all
> is not an option.
> ------------------------------
>
> In lib/ansible/modules/database/postgresql/postgresql_query.py
> <https://github.com/ansible/ansible/pull/25411#discussion_r128397703>:
>
> > +     PostgreSQL must also be installed on the remote host. For Ubuntu-based
> +     systems, install the postgresql, libpq-dev, and python-psycopg2 packages
> +     on the remote host before using this module.
> +   - If the passlib library is installed, then passwords that are encrypted
> +     in the DB but not encrypted when passed as arguments can be checked for
> +     changes. If the passlib library is not installed, unencrypted passwords
> +     stored in the DB encrypted will be assumed to have changed.
> +   - If you specify PUBLIC as the user, then the privilege changes will apply
> +     to all users. You may not specify password or role_attr_flags when the
> +     PUBLIC user is specified.
> +   - The ssl_rootcert parameter requires at least Postgres version 8.4 and I(psycopg2) version 2.4.3.
> +requirements: [ psycopg2 ]
> +author:
> +    - "Felix Archambault (@archf)"
> +    - "Will Rouesnel (@wrouesnel)"
> +'''
>
> Existing documentation should be reused not duplicated. Reuse existing documentation
> fragment
> <http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#documentation-fragments>
> :
>
> extends_documentation_fragment:
> - postgres
>
> ------------------------------
>
> In lib/ansible/modules/database/postgresql/postgresql_query.py
> <https://github.com/ansible/ansible/pull/25411#discussion_r128397943>:
>
> > +module: postgresql_query
> +short_description: Run arbitrary queries on a postgresql database.
> +description:
> +   - Run arbitrary SQL queries against a PostgreSQL instance from the current host.
> +   - Query results are returned as "query_results" as a list of dictionaries.
> +     Number of rows affected are returned as "row_count".
> +   - Can read queries from a .sql script files.
> +   - SQL scripts may be templated with ansible variables at execution time.
> +   - Run queries in check_mode (will roll them back, autocommit must be disabled). This may affect performance of a busy cluster and it may incur side effects.
> +version_added: "2.4"
> +options:
> +  db:
> +    description:
> +      - name of the database to run queries against.
> +    required: false
> +    default: postgres
>
> Default value for db parameter isn't enforced in argument_spec: I would
> rather remove this line or add a default value for db in argument_spec.
> ------------------------------
>
> In lib/ansible/modules/database/postgresql/postgresql_query.py
> <https://github.com/ansible/ansible/pull/25411#discussion_r128399112>:
>
> > +    required: false
> +    default: localhost
> +  login_unix_socket:
> +    description:
> +      - Path to a Unix domain socket for local connections.
> +    required: false
> +    default: null
> +  query:
> +    description:
> +      - SQL query to run. Variables can be escaped with psycopg2 syntax.
> +      - Can be a sql scripts file.
> +    required: true
> +  positional_args:
> +    description:
> +      - A list of values to be passed as positional arguments to the query.
> +      - Cannot be used with named_args.
>
> Use formatting function
> <http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#formatting-options>
> when referring to other options.
> ------------------------------
>
> In lib/ansible/modules/database/postgresql/postgresql_query.py
> <https://github.com/ansible/ansible/pull/25411#discussion_r128399122>:
>
> > +    required: false
> +    default: null
> +  login_host:
> +    description:
> +      - Host running PostgreSQL.
> +    required: false
> +    default: localhost
> +  login_unix_socket:
> +    description:
> +      - Path to a Unix domain socket for local connections.
> +    required: false
> +    default: null
> +  query:
> +    description:
> +      - SQL query to run. Variables can be escaped with psycopg2 syntax.
> +      - Can be a sql scripts file.
>
> Another parameter dedicated to SQL script could be used (for example
> script). Then query and script would be mutually exclusive. The type of
> script would be path, the main benefit is that the value would be expanded
> <https://github.com/archf/ansible/blob/e7958745d7c9812d0c1cd64de09be92e1f302e09/lib/ansible/module_utils/basic.py#L1801>
> using os.path.expanduser and os.path.expandvars.
>
> s/sql/SQL/
> ------------------------------
>
> In lib/ansible/modules/database/postgresql/postgresql_query.py
> <https://github.com/ansible/ansible/pull/25411#discussion_r128399330>:
>
> > +      - Password used to authenticate with PostgreSQL.
> +    required: false
> +    default: null
> +  login_host:
> +    description:
> +      - Host running PostgreSQL.
> +    required: false
> +    default: localhost
> +  login_unix_socket:
> +    description:
> +      - Path to a Unix domain socket for local connections.
> +    required: false
> +    default: null
> +  query:
> +    description:
> +      - SQL query to run. Variables can be escaped with psycopg2 syntax.
>
> Add a link to Psycopg documentation using U formatting function.
> ------------------------------
>
> In lib/ansible/modules/database/postgresql/postgresql_query.py
> <https://github.com/ansible/ansible/pull/25411#discussion_r128399865>:
>
> > +    db: acme
> +    login_user: django
> +    login_password: ceec4eif7ya
> +    query: "{{playbook_dir}}/scripts/my_sql_query_file.sql"
> +    named_args:
> +      a_value: "positional string value 1"
> +      b_value: "positional string value 2"
> +
> +# Run queries from a '.sql' file and assign result in a fact available for the
> +# rest of the ansible runtime. Query inside scripts may contain 'named_args'
> +# or 'positional_args'.
> +- postgresql_query:
> +    db: acme
> +    login_user: django
> +    login_password: ceec4eif7ya
> +    query: SELECT * FROM some_table WHERE a_column=%(a_value)s AND b_column=%(b_value)s
>
> Remove the first query occurrence.
> ------------------------------
>
> In lib/ansible/modules/database/postgresql/postgresql_query.py
> <https://github.com/ansible/ansible/pull/25411#discussion_r128400081>:
>
> > +'''
> +
> +RETURN = '''
> +query_results:
> +    description: List of dictionaries in column:value form representing returned rows.
> +    returned: changed
> +    type: list
> +    sample: [{"Column": "Value1"},{"Column": "Value2"}]
> +row_count:
> +    description: number of affected rows by query, if applicable
> +    returned: changed
> +    type: int
> +    sample: 5
> +'''
> +
> +HAS_PSYCOPG2 = False
>
> HAS_PSYCOPG2 is already defined in ansible.module_utils.postgres which is
> imported below. Remove this block.
> ------------------------------
>
> In lib/ansible/modules/database/postgresql/postgresql_query.py
> <https://github.com/ansible/ansible/pull/25411#discussion_r128400428>:
>
> > +    import psycopg2.extras
> +except ImportError:
> +    pass
> +else:
> +    HAS_PSYCOPG2 = True
> +
> +import traceback
> +from ansible.module_utils.six import iteritems
> +from ansible.module_utils.basic import AnsibleModule, get_exception, BOOLEANS
> +import ansible.module_utils.postgres as pgutils
> +
> +
> +def main():
> +    argument_spec = pgutils.postgres_common_argument_spec()
> +
> +    argument_spec.update(dict(
>
> default value of default is None, there is no need to specify default=None.
> default value of type is str, there is no need to use type="str".
> ------------------------------
>
> In lib/ansible/modules/database/postgresql/postgresql_query.py
> <https://github.com/ansible/ansible/pull/25411#discussion_r128400513>:
>
> > +        query=dict(type="str"),
> +        autocommit=dict(type='bool', default=False, choices=BOOLEANS),
> +        positional_args=dict(type="list"),
> +        named_args=dict(type="dict"),
> +        fact=dict(default=None),
> +    ))
> +
> +    module = AnsibleModule(
> +        argument_spec=argument_spec,
> +        mutually_exclusive=[
> +            ["positional_args", "named_args"]
> +        ],
> +        supports_check_mode=True,
> +    )
> +
> +    if not HAS_PSYCOPG2:
>
> pgutils.ensure_libs() is used below, there is no need to check
> HAS_PSYCOPG2 here.
> ------------------------------
>
> In lib/ansible/modules/database/postgresql/postgresql_query.py
> <https://github.com/ansible/ansible/pull/25411#discussion_r128401097>:
>
> > +            else:
> +                db_connection.set_isolation_level(psycopg2.extensions.ISOLATION_LEVEL_AUTOCOMMIT)
> +
> +        # Using RealDictCursor allows access to row results by real column name
> +        cursor = db_connection.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
> +
> +    except TypeError:
> +        e = get_exception()
> +        if 'sslrootcert' in e.args[0]:
> +            module.fail_json(msg='''Postgresql server must be at least version
> +                             8.4 to support sslrootcert.
> +                             Exception: {0}'''.format(e), exception=traceback.format_exc())
> +        module.fail_json(msg="unable to connect to database: %s" % e, exception=traceback.format_exc())
> +
> +    except Exception:
> +        e = get_exception()
>
> get_exception was used for Python 2.4 compatibility, support for 2.4 on
> the managed system's side was dropped: use native Python exception handling.
> ------------------------------
>
> In lib/ansible/modules/database/postgresql/postgresql_query.py
> <https://github.com/ansible/ansible/pull/25411#discussion_r128402158>:
>
> > +    statusmessage = cursor.statusmessage
> +
> +    changed = False
> +
> +    # set changed flag only on non read-only command
> +    if ("UPDATE" in statusmessage or "INSERT" in statusmessage or
> +            "UPSERT" in statusmessage or "DELETE" in statusmessage):
> +        changed = True
> +
> +    if changed:
> +        if module.check_mode:
> +            db_connection.rollback()
> +        else:
> +            db_connection.commit()
> +
> +    db_connection.close()
>
> db_connection.close() is not called when cursor.execute raises an
> exception.
> ------------------------------
>
> In lib/ansible/modules/database/postgresql/postgresql_query.py
> <https://github.com/ansible/ansible/pull/25411#discussion_r128402533>:
>
> > +    # prepare args
> +    if module.params["positional_args"] is not None:
> +        arguments = module.params["positional_args"]
> +
> +    elif module.params["named_args"] is not None:
> +        arguments = module.params["named_args"]
> +
> +    try:
> +        cursor.execute(query, arguments)
> +    except Exception:
> +        e = get_exception()
> +        module.fail_json(msg="Unable to execute query: %s" % e,
> +                         query_arguments=arguments)
> +
> +    ansible_facts = {}
> +    query_results = []
>
> rowcount could be set to 0 then the else block of the conditional
> statement could be removed.
> ------------------------------
>
> In lib/ansible/modules/database/postgresql/postgresql_query.py
> <https://github.com/ansible/ansible/pull/25411#discussion_r128405135>:
>
> > +
> +        rowcount = len(query_results)
> +        fact = module.params["fact"]
> +        if fact is not None:
> +            ansible_facts = {fact: query_results}
> +    else:
> +        rowcount = 0
> +
> +    statusmessage = cursor.statusmessage
> +
> +    changed = False
> +
> +    # set changed flag only on non read-only command
> +    if ("UPDATE" in statusmessage or "INSERT" in statusmessage or
> +            "UPSERT" in statusmessage or "DELETE" in statusmessage):
> +        changed = True
>
> What about MOVE, COPY, CREATE ? It doesn't seem reliable, changed should
> be True if the query has been executed. By the way I am not sure if
> supporting check_mode is a good idea. command and shell modules don't
> support check_mode, it would be fine it this module didn't support
> check_mode.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ansible/ansible/pull/25411#pullrequestreview-51072727>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGSSfJyXsJaeYnrhYVuX6cTQXpdscqrcks5sPqlTgaJpZM4NyGka>
> .
>

pilou- commented:
@archf just for you to know, Ansible 2.4 feature freeze for community module is planned for 29 August 2017.

wrouesnel commented:
@archf I think based on @pilou- comments that the check_mode should be removed for the time being. Without a truly comprehensive test suite I feel like the risk of it doing something it shouldn't is just too high (and ergo, is making getting this merged trickier).

wrouesnel commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,347 @@
+    argument_spec.update(dict(
+        db=dict(default=None),
+        query=dict(type="str"),
+        autocommit=dict(type='bool', default=False, choices=BOOLEANS),
```
I needed to patch this line on the latest devel head to this in order to let the module work.
```
        autocommit=dict(type='bool', default=False, choices=tuple(BOOLEANS)),
```

pilou- replied:
Quoting [dev_guide](/docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html):
> If an argument is type='bool', this field should be set to type: bool and no choices should be specified.

archf committed:
postgresql_query_module: fixes and improvements

* db defaults to None (same as psql)
* more code reuse from pgutils import
* graceful connection closing in every catched exception
* dropped python 2.4 get_exception() when catching exception
* initialize rowcount to remove else statement
* grouped variable initialization on top
* report 'changed' on more statusmessages
* fix typo s/stout_lines/stdout_lines
* RealDictCursor objects vs python dict (as mentionned by wrouesnel)
* disable check mode

archf commented:
Thanks a lot @pilou- and @wrouesnel for you review. I made a first patch from your comments. I'll try to cover remaining stuff before the weekend (module args & documentation fragments)...

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,330 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# This file is part of Ansible
```
Use the [short version](http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#copyright):
```
#!/usr/bin/python
# Copyright (c) 2017 Felix Archambault
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

just below add (here is an [example](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/database/postgresql/postgresql_db.py#L6)):
```
from __future__ import absolute_import, division, print_function
__metaclass__ = type
```

pilou- replied:
Done

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,330 @@
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+ANSIBLE_METADATA = {'metadata_version': '1.0',
```
update `metadata_version` to `1.1` (nothing else need to be updated). [Metadata documentation](http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#ansible-metadata-block) isn't up to date, see https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/dev_guide/developing_modules_documenting.rst#version-1-1-of-the-metadata.

pilou- replied:
Updated

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,330 @@
+  ssl_mode:
+    description:
+      - Determines whether or with what priority a secure SSL TCP/IP connection will be negotiated with the server.
+      - See https://www.postgresql.org/docs/current/static/libpq-ssl.html for more information on the modes.
```
Use `U()` [formatting function](http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#formatting-options) around URL ([sample](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/lxd/lxd_profile.py#L32)).

pilou- replied:
Fixed (doc fragment used)

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,330 @@
+              if k in params_map and v != '' and v is not None)
+
+    # If a login_unix_socket is specified, incorporate it here.
+    is_localhost = "host" not in kw or kw["host"] == "" or kw["host"] == "localhost"
```
This doesn't seems right:
- `kw["host"]` can not be equal to empty string
- What if user specified `127.0.0.1` ?

`login_unix_socket` parameter should be removed because `host` could be used instead but then documentation fragment and `module_utils.postgres` could not be used :-/ I propose to keep `login_unix_socket` remove `is_localhost` and add `['host', 'login_unix_socket']` to `mutually_exclusive`.

[Connection URIs](https://www.postgresql.org/docs/current/static/libpq-connect.html#LIBPQ-CONNSTRING) should be used, it should be done for all PostgreSQL modules. Something to do for Ansible 2.5 :)

pilou- replied:
> I propose to keep login_unix_socket remove is_localhost and add ['host', 'login_unix_socket'] to mutually_exclusive.

must be done for this module. Other improvements should be solved in another pull request for all PostgreSQL modules.

archf replied:
Your solution is much neater. I removed the is_localhost thing i shamely copied from other prostgres modules...

Yes URI, one day... Would be saner...

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,330 @@
+
+    except TypeError as e:
+        if 'sslrootcert' in e.args[0]:
+            module.fail_json(msg='''Postgresql server must be at least version
```
s/Postgresql/PostgreSQL/

archf replied:
Fixed

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,330 @@
+        # Using RealDictCursor allows access to row results by real column name
+        cursor = db_connection.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
+
+    except TypeError as e:
```
- `TypeError` won't be raised: `ensure_libs` raise `LibraryError`.
- There is no need to differentiate `LibraryError` and `Exception`. 
- `db_connection.close()` must called before `module.fail_json()` (`fail_json` calls `sys.exit` nothing is executed after)

 It seems the `try` block could be updated:
```
try:
    ensure_libs(...)
    [...]
except Exception as e:
    db_connection.close()
    module.fail_json(msg=to_native(e), exception=traceback.format_exc())
```

archf replied:
I think `TypeError` can be catched as it is placed before `Exception` that is less specific. 

I agree there is no need to differentiate `LibraryError` and `Exception`. I think we need to leave this be for the time beeing inside `postgres.py` until we harmonize all postgres modules further more. i fixed the `db_connection.close()` already.

pilou- replied:
Two things:
- `ensure_libs(...)` raise `LibraryError` when `sslrootcerts` isn't supported: should not `TypeError` be replaced with `LibraryError` ? If not which statement might raise a `TypeError` exception ?
- `ensure_libs()` raises an error telling that [_psycopg2 must be at least 2.4.3_](https://github.com/archf/ansible/blob/bf17b7df2db50ac4c91816ec3737c0c2c35dc994/lib/ansible/module_utils/postgres.py#L45) and this module states that `Postgresql server must be at least version 8.4 to support sslrootcert`. In order to use [`sslrootcert`](https://www.postgresql.org/docs/8.4/static/libpq-connect.html#LIBPQ-CONNECT-SSLROOTCERT), `libpq` version should be at least 8.4 but PostgreSQL could be older. `sslrootcert` feature doesn't require a minimum PostgreSQL server version. It seems exception message raised by `ensure_libs` should be `sslrootcert: psycopg2 >= 2.4.3 and libpq >= 8.4 required` (it should be fixed in another pull request). That's why I would prefer only one `except` block there.

archf replied:
ok, i removed it. Postgresql 8.X is not something i really care about. I was just not brave enough to try install Postgres that old and fiddle with it so i left it there and did not question to much. Might be on RHEL 5.X... Anyhow we are still catching an exception...

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,330 @@
+        try:
+            query = open(query, 'r').read().strip('\n')
+        except Exception as e:
+            module.fail_json(msg="Unable to find '%s' in given path." % query)
```
`db_connection.close()` must be called before `fail_json`.

pilou- replied:
Fixed

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,330 @@
+     PostgreSQL must also be installed on the remote host. For Ubuntu-based
+     systems, install the postgresql, libpq-dev, and python-psycopg2 packages
+     on the remote host before using this module.
+   - If the passlib library is installed, then passwords that are encrypted
```
The sentences about `passlib` and `role_attr_flags` seems irrelevant.

pilou- replied:
Fixed (doc fragment used)

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,330 @@
+   - If you specify PUBLIC as the user, then the privilege changes will apply
+     to all users. You may not specify password or role_attr_flags when the
+     PUBLIC user is specified.
+   - The ssl_rootcert parameter requires at least Postgres version 8.4 and I(psycopg2) version 2.4.3.
```
s/Postgres/PostgreSQL/

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,330 @@
+    try:
+        cursor.execute(query, arguments)
+    except Exception as e:
+        module.fail_json(msg="Unable to execute query: %s" % e,
```
`db_connection.close()` must be called before `fail_json`.

pilou- replied:
Fixed

archf committed:
postgresql_query_module: improve module args and documentation

* fix db connection closure
* use postgresql fragments
* use argument 'name' instead of 'fact' (same as include_vars)
* removed booleans choices
* unset the query arg type and make it mandatory

archf committed:
improve pathing support

expand paths using os.path.expanduser

Signed-off-by: Félix Archambault <fel.archambault@gmail.com>

archf commented:
Some more update. Maybe as @pilou- suggested we should have a separate arg for script file. I'd like the thing to look in the role `files` or `scripts` directory. Not sure how to do that thought.

I feel like the other things such as the args and some exceptions should be solved in all postgresql modules at the same time by improving pgutils as bit. I'll need to sleep on this...

archf committed:
added __future__ imports

I do not use them and this seems like a waste of resources but i add
them for the bot to stop complaining.

Signed-off-by: Félix Archambault <fel.archambault@gmail.com>

pilou- commented:
These comments need to be addressed:
- https://github.com/ansible/ansible/pull/25411#discussion_r134716767
- https://github.com/ansible/ansible/pull/25411#discussion_r134714645

_Both comments have been addressed_.

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_query.py:
```diff
@@ -0,0 +1,275 @@
+        default: None
+    query:
+        description:
+            - SQL query to run. Variables can be escaped with psycopg2 syntax.
```
A link to pyscopg2 documentation could be added (using `U()` formatting function).

archf replied:
I added a link. Can be handy indeed.

archf committed:
postgresql_query_module: added url in query argument description

url to psycopg documentation

archf committed:
postgresql_query_module: fix login_host and login_unix_socket exclusivity

archf committed:
postgresql_query_module: remove TypeError exception

archf committed:
postgresql_query_module: Fix the module examples that were wrong

archf committed:
use login_unix_socket if received

assigned its value to 'host'
see: http://initd.org/psycopg/docs/module.html#psycopg2.connect

archf committed:
postgresql_query_module: bump version to 2.5 since 2.4 release was missed

archf committed:
postgresql_query_module: connection parameter: database is a deprecated alias

pilou- commented:
@arch in order to remove the [`needs_revision`](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md#label-needs_revision) label, you should add a [`ready_for_review`](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md#cmd-ready_for_review) comment.

@wrouesnel  @Akasurde this pull request has been updated, could you review it again ?

archf commented:
ok, doing that. Thanks again! I think we reached something much better...

archf commented:
`ready_for_review`

japsu commented:
I propose we support the use case of **ensuring some row exists** in the database, such as creating an initial user for an application. Something like this:

    - postgresql_query:
        query: INSERT INTO users (username, password_hash) VALUES (…, …)
        only_if_not_exists: SELECT id FROM users WHERE username = 'joeadmin'

What should the "run this query first and only if it returns and empty set, run the actual query" parameter be called? `only_if_not_exists`? `test_query`? `ok_when`?

The semantics should be that the query reports `ok` (not `changed`) iff the `only_if_not_exists` query returns a non-empty set.

Might warrant `state: present|absent` as well (perhaps a `remove_query` for `absent`).

pilou- commented:
As I wrote before [tests must be added](https://github.com/ansible/ansible/pull/25411#pullrequestreview-51072727). There are [existing integration tests for other PostgreSQL modules](https://github.com/ansible/ansible/tree/devel/test/integration/targets/postgresql/tasks), a new file (something like `test_postgresql_query.yml`) should be added there.

archf commented:
I'll add tests then.

On Feb 7, 2018 17:13, "Pilou" <notifications@github.com> wrote:

> As I wrote before tests must be added
> <https://github.com/ansible/ansible/pull/25411#pullrequestreview-51072727>.
> There are existing integration tests for other PostgreSQL modules
> <https://github.com/ansible/ansible/tree/devel/test/integration/targets/postgresql/tasks>,
> a new file (something like test_postgresql_query.yml) should be added
> there.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ansible/ansible/pull/25411#issuecomment-363928801>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGSSfOcLzqnbQGiHKNteFKMUR1VTso-Tks5tSiAdgaJpZM4NyGka>
> .
>

sebasmannem commented:
@archf : I needed to do the same for the pg_hba module.
If you like, I can help.

archf commented:
Hi @sebasmannem! By all mean go ahead. I promised myself to finish this for many months but reality is I won't be able in my current context to work on it at least  for the first half of the year.

Andersson007 commented:
@archf the idea of this module is simple and good. If you don't have enough time we can rewrite it and create a new pull request instead (keeping info about current authors, of course)

archf commented:
Go ahead by all mean! I won't stop you!

Andersson007 commented:
@archf I'm going to complete this (or maybe anybody else) when we write integration tests for the existing modules.

Andersson007 commented:
I completely rewrote this module.
Now it's here https://github.com/ansible/ansible/pull/52555
@archf , please, close this PR.
I added you and @wrouesnel as co-authors.
It will be great if you review https://github.com/ansible/ansible/pull/52555 module and approve or disapprove it.
